targets/default/jerry-port-default-jobqueue.c

```diff
@@ -0,0 +1,122 @@
+/* Copyright JS Foundation and other contributors, http://js.foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "jerry-port.h"
+#include "jerry-port-default.h"
+#include "ecma-globals.h"
+#include "jmem.h"
+#include "ecma-jobqueue.h"
+
+#ifndef CONFIG_DISABLE_ES2015_PROMISE_BUILTIN
+
+typedef struct jerry_port_queueitem_t jerry_port_queueitem_t;
+
+/**
+ * Description of the queue item.
+ */
+struct jerry_port_queueitem_t
+{
+  jerry_port_queueitem_t *next_p; /**< point to next item */
+  ecma_job_t *job_p; /**< point to the job */
+};
+
+/**
+ * Description of a job queue (FIFO)
+ */
+typedef struct
+{
+  jerry_port_queueitem_t *head_p; /**< point to the head item of the queue */
+  jerry_port_queueitem_t *tail_p; /**< point to the tail item of the queue*/
+} jerry_port_jobqueue_t;
+
+jerry_port_jobqueue_t queue;
+
+/**
+ * Initialize the job queue
+ */
+void jerry_port_jobqueue_init (void)
+{
+  queue.head_p = NULL;
+  queue.tail_p = NULL;
+} /* jerry_port_jobqueue_init */
+
+/**
+ * Enqueue a job
+ */
+void jerry_port_jobqueue_enqueue (ecma_job_t *job_p) /**< the ecma job */
+{
+  jerry_port_queueitem_t *item_p = jmem_heap_alloc_block (sizeof (jerry_port_queueitem_t));
+  item_p->job_p = job_p;
+
+  if (queue.head_p == NULL)
+  {
+    JERRY_ASSERT (queue.tail_p == NULL);
+
+    queue.head_p = item_p;
+    item_p->next_p = NULL;
+    queue.tail_p = queue.head_p;
+
+    return;
+  }
+
+  JERRY_ASSERT (queue.tail_p != NULL);
+
+  queue.tail_p->next_p = item_p;
+  queue.tail_p = item_p;
+} /* jerry_port_jobqueue_enqueue */
+
+/**
+ * Dequeue and get the job.
+ */
+static ecma_job_t *
+jerry_port_jobqueue_dequeue (void)
+{
+  if (queue.head_p == NULL)
+  {
+    JERRY_ASSERT (queue.tail_p == NULL);
+
+    return NULL;
+  }
+
+  JERRY_ASSERT (queue.tail_p != NULL);
+
+  jerry_port_queueitem_t *item_p = queue.head_p;
+  queue.head_p = queue.head_p->next_p;
+  ecma_job_t *job_p = item_p->job_p;
+
+  jmem_heap_free_block (item_p, sizeof (jerry_port_queueitem_t));
+
+  return job_p;
+} /* jerry_port_jobqueue_dequeue */
+
+/**
+ * Start the jobqueue.
+ */
+void jerry_port_jobqueue_run (void)
+{
+  while (1)
+  {
+    ecma_job_t *job_p = jerry_port_jobqueue_dequeue ();
+
+    if (job_p == NULL)
+    {
+      return;
+    }
+
+    ecma_job_process (job_p);
```

Comment:
`ecma_job_process` should probably be exposed in `jerry-port.h` somehow, to avoid reaching into internals.

Hmm, this is the first time the port needs to call back into JrS, I think?

Is there a convention to indicate in a port header that a function declaration is there to indicate the port can/needs to call it? (I'm used to adding `extern` as a hint, but don't know if that's a widespread convention or not.)